internal/experiment/experiment.go

```diff
@@ -0,0 +1,137 @@
+// By default experiment mode is disabled, but when enabled, experimental features can be enabled.
+// These features are not yet stable and may change in the future.
+//
+// Note that any behavior outlined here should be documented in /docs/_docs/04_reference/experiment-mode.md
+//
+// That is how users will know what to expect when they enable experiment mode, and how to customize it.
+package experiment
+
+import (
+	"strings"
+)
+
+// NewExperiments returns a new Experiments map with all experiments disabled.
+//
+// Bottom values for each experiment are the defaults, so only the names of experiments need to be set.
+func NewExperiments() Experiments {
+	return Experiments{
+		Symlinks: Experiment{
+			Name: Symlinks,
+		},
+	}
+}
+
+// Experiment represents an experiment that can be enabled.
+// When the experiment is enabled, Terragrunt will behave in a way that uses some experimental functionality.
+type Experiment struct {
+	// Enabled determines if the experiment is enabled.
+	Enabled bool
+	// Name is the name of the experiment.
+	Name string
+	// Status is the status of the experiment.
+	Status int
+}
+
+func (e Experiment) String() string {
+	return e.Name
+}
+
+const (
+	// Symlinks is the experiment that allows symlinks to be used in Terragrunt configurations.
+	Symlinks = "symlinks"
+)
+
+const (
+	// StatusOngoing is the status of an experiment that is ongoing.
+	StatusOngoing = iota
+	// StatusCompleted is the status of an experiment that is completed.
+	StatusCompleted
+)
+
+type Experiments map[string]Experiment
+
+// ValidateExperimentNames validates the given slice of experiment names are valid.
+func (e *Experiments) ValidateExperimentNames(experimentNames []string) (string, error) {
+	completedExperiments := []string{}
+	invalidExperiments := []string{}
+
+	for _, name := range experimentNames {
+		experiment, ok := (*e)[name]
+		if !ok {
+			invalidExperiments = append(invalidExperiments, name)
+			continue
+		}
+
+		if experiment.Status == StatusCompleted {
+			completedExperiments = append(completedExperiments, name)
+		}
+	}
+
+	var warning string
+	if len(completedExperiments) > 0 {
+		warning = CompletedExperimentsWarning{
+			ExperimentNames: completedExperiments,
+		}.String()
+	}
+
+	var err error
+	if len(invalidExperiments) > 0 {
+		err = InvalidExperimentsError{
+			ExperimentNames: invalidExperiments,
+		}
+	}
+
+	return warning, err
+}
+
+// EnableExperiments enables the given experiments.
+func (e *Experiments) EnableExperiments(experimentNames []string) error {
+	invalidExperiments := []string{}
+
+	for _, name := range experimentNames {
+		experiment, ok := (*e)[name]
+		if !ok {
+			invalidExperiments = append(invalidExperiments, name)
+			continue
+		}
+
+		experiment.Enabled = true
+		(*e)[name] = experiment
+	}
+
+	if len(invalidExperiments) > 0 {
+		return InvalidExperimentsError{
+			ExperimentNames: invalidExperiments,
+		}
```

Comment:
```suggestion
		return errors.New(InvalidExperimentsError{
			ExperimentNames: invalidExperiments,
		})
```

to have a stack trace.